dht.go

```diff
@@ -365,6 +375,55 @@ func (dht *IpfsDHT) betterPeersToQuery(pmes *pb.Message, p peer.ID, count int) [
 	return filtered
 }
 
+func (dht *IpfsDHT) SetMode(m int) error {
+	dht.modeLk.Lock()
+	defer dht.modeLk.Unlock()
+
+	if m == dht.mode {
+		return nil
+	}
+
+	switch m {
+	case ModeServer:
+		return dht.moveToServerMode()
+	case ModeClient:
+		return dht.moveToClientMode()
+	default:
+		return fmt.Errorf("unrecognized dht mode: %d", m)
+	}
+}
+
+func (dht *IpfsDHT) moveToServerMode() error {
+	dht.mode = ModeServer
+	for _, p := range dht.protocols {
+		dht.host.SetStreamHandler(p, dht.handleNewStream)
+	}
+	return nil
+}
+
+func (dht *IpfsDHT) moveToClientMode() error {
+	dht.mode = ModeClient
+	for _, p := range dht.protocols {
+		dht.host.RemoveStreamHandler(p)
+	}
+
+	pset := make(map[protocol.ID]bool)
+	for _, p := range dht.protocols {
+		pset[p] = true
+	}
+
+	// hacky... also closes both inbound and outbound streams
+	for _, c := range dht.host.Network().Conns() {
+		for _, s := range c.GetStreams() {
+			if pset[s.Protocol()] {
+				// is it even safe to call close concurrently with a read/write call?
+				s.Close()
```

Comment:
we might interfere with ongoing queries here, we probably want to keep the outbound streams open.

Reply:
I don't think we have a way to tell direction from here though

Reply:
close should only half close anyway, i think that's okay

Reply:
* We'll want to reset these. That should be safe.
* `Stat()` includes a direction.

Can we just record all inbound streams in a map somewhere? This would _also_ help us cleanup on close.

Reply:
You could maybe try intersecting with the streams inside the _messageSender_s. I think those are the outbound streams.

We could also close the connection if it contains no more streams (unlikely though). That would force older peers to drop us from their table.

Reply:
ah yes, `Stat()` was one of my first changes :) that should do it.